frontend/src/components/widgets/DataFrame/columns/cells/DateTimeCell.tsx

```diff
@@ -0,0 +1,209 @@
+    case "date":
+      return isoDate.split("T")[0]
+    case "datetime-local":
+      return isoDate.replace("Z", "")
+    case "time":
+      return isoDate.split("T")[1].replace("Z", "")
+    default:
+      throw new Error(`Unknown date kind ${dateKind}`)
+  }
+}
+
+export type DateTimeCell = CustomCell<DateTimeCellProps>
+
+const Editor: ReturnType<ProvideEditorCallback<DateTimeCell>> = cell => {
+  const cellData = cell.value.data
+  const { format, displayDate } = cellData
+
+  const step =
+    cellData.step !== undefined && !Number.isNaN(Number(cellData.step))
+      ? Number(cellData.step)
+      : undefined
+
+  const minValue =
+    cellData.min instanceof Date
+      ? formatValueForHTMLInput(format, cellData.min)
+      : cellData.min
+
+  const maxValue =
+    cellData.max instanceof Date
+      ? formatValueForHTMLInput(format, cellData.max)
+      : cellData.max
+
+  let date = cellData.date
+  // Convert timezone offset to milliseconds
+  const timezoneOffsetMs = cellData.timezoneOffset
+    ? cellData.timezoneOffset * 60 * 1000
+    : 0
+  if (timezoneOffsetMs && date) {
+    // Adjust based on the timezone offset
+    date = new Date(date.getTime() + timezoneOffsetMs)
+  }
+  const value = formatValueForHTMLInput(format, date)
+  if (cell.value.readonly) {
+    return (
+      <TextCellEntry
+        highlight={true}
+        autoFocus={false}
+        disabled={true}
+        value={displayDate ?? ""}
+        onChange={() => undefined}
+      />
+    )
+  }
+
+  return (
+    <StyledInputBox
+      data-testid={"date-time-cell"}
+      required
+      type={format}
+      defaultValue={value}
+      min={minValue}
+      max={maxValue}
+      step={step}
+      autoFocus={true}
+      onChange={event => {
+        if (isNaN(event.target.valueAsNumber)) {
+          // The user has cleared the date, contribute as undefined
+          cell.onChange({
+            ...cell.value,
+            data: {
+              ...cell.value.data,
+              date: undefined,
+            },
+          })
+        } else {
+          cell.onChange({
+            ...cell.value,
+            data: {
+              ...cell.value.data,
+              // use valueAsNumber because valueAsDate is null for "datetime-local"
+              // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#technical_summary
+              date: new Date(event.target.valueAsNumber - timezoneOffsetMs),
+            },
+          })
+        }
+      }}
+    />
+  )
+}
+
+export const DateTimeCellRenderer: CustomRenderer<DateTimeCell> = {
+  kind: GridCellKind.Custom,
+  isMatch: (cell: CustomCell): cell is DateTimeCell =>
+    (cell.data as any).kind === "date-time-cell",
+  draw: (args, cell) => {
+    const { displayDate } = cell.data
+    drawTextCell(args, displayDate, cell.contentAlign)
+    return true
+  },
+  measure: (ctx, cell) => {
+    const { displayDate } = cell.data
+    return ctx.measureText(displayDate).width + 16
```

Comment:
Should add a comment for this +16 it feels like. Feels like an odd width addition.

Reply:
Added a comment